x/ecocredit/server/marketplace/buy.go

```diff
@@ -0,0 +1,241 @@
+			if !ok {
+				return nil, fmt.Errorf("could not convert sell order's ask price to %T: %s", sdk.Int{}, sellOrder.AskPrice)
+			}
+			if order.BidPrice.Amount.LT(askAmount) {
+				return nil, sdkerrors.ErrInsufficientFunds.Wrapf("bid price too low: got %s, needed at least %s",
+					order.BidPrice.Amount.String(), sellOrder.AskPrice)
+			}
+
+			if err = k.updateBalances(ctx, sellOrder, buyerAcc, orderAmt, *order.BidPrice); err != nil {
+				return nil, fmt.Errorf("error updating balances: %w", err)
+			}
+			if !sellOrder.DisableAutoRetire {
+				if err = sdkCtx.EventManager().EmitTypedEvent(&core.EventRetire{
+					Retirer:    buyerAcc.String(),
+					BatchDenom: batch.BatchDenom,
+					Amount:     order.Quantity,
+					Location:   order.RetirementLocation,
+				}); err != nil {
+					return nil, err
+				}
+			} else {
+				if err = sdkCtx.EventManager().EmitTypedEvent(&core.EventReceive{
+					Sender:         sdk.AccAddress(sellOrder.Seller).String(),
+					Recipient:      buyerAcc.String(),
+					BatchDenom:     batch.BatchDenom,
+					TradableAmount: order.Quantity,
+					RetiredAmount:  "",
+					BasketDenom:    "",
+				}); err != nil {
+					return nil, err
+				}
+			}
+		default:
+			return nil, sdkerrors.ErrInvalidRequest.Wrap("only direct buy orders are enabled at this time")
+		}
+	}
+	return &v1.MsgBuyResponse{}, nil
+}
+
+// updateBalances moves credits according to the order. it will:
+// - update a sell order, removing it if quantity becomes 0 as a result of this purchase.
+// - remove the purchaseQty from the seller's escrowed balance.
+// - add credits to the buyer's tradable/retired address (based on the DisableAutoRetire field).
+// - update the supply accordingly.
+// - send the coins specified in the bid to the seller.
+func (k Keeper) updateBalances(ctx context.Context, sellOrder *api.SellOrder, buyerAcc sdk.AccAddress, purchaseQty math.Dec, bidCoin sdk.Coin) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	sellOrderQty, err := math.NewDecFromString(sellOrder.Quantity)
+	if err != nil {
+		return err
+	}
+	// since we only support direct buy orders at this time, we fail when the requested purchase amount is more than
+	// the available credits for sale, rather than partial filling the buy order.
+	sellOrderQty, err = math.SafeSubBalance(sellOrderQty, purchaseQty)
+	if err != nil {
+		return err
+	}
+	if sellOrderQty.IsZero() { // remove the sell order if no credits are left
+		if err := k.stateStore.SellOrderTable().Delete(ctx, sellOrder); err != nil {
+			return err
+		}
+	} else { // update the sell order with the new value otherwise
+		sellOrder.Quantity = sellOrderQty.String()
+		if err = k.stateStore.SellOrderTable().Update(ctx, sellOrder); err != nil {
+			return err
+		}
+	}
+
+	// remove the credits from the seller's escrowed balance
+	sellerBal, err := k.coreStore.BatchBalanceTable().Get(ctx, sellOrder.Seller, sellOrder.BatchId)
+	if err != nil {
+		return err
+	}
```

Comment:
You formatted the error with other get calls above. Would be nice to do the same here.

Reply:
I'm noticing more errors below that would be nice to format to include more information. I realize all these errors are already formatted with "error updating balances" but I think we can provide more context.